app/services/cart.js

```diff
@@ -1,179 +1,382 @@
+    );
+  }),
 
-    if (operation == "delete") {
-      return this.setItemAvailability(cartItem, false);
+  isLoggedIn: Ember.computed.alias("session.authToken"),
+  isEmpty: Ember.computed.empty("cartItems"),
+  isNotEmpty: Ember.computed.not("isEmpty"),
+  counter: Ember.computed.alias("cartItems.length"),
+  user: Ember.computed.alias("session.currentUser"),
+  userId: Ember.computed.alias("user.id"),
+
+  // ---- Hooks
+
+  onUserChange() {
+    const loggedIn = this.get("session.authToken");
+    if (loggedIn) {
+      this.onLogin();
+    } else {
+      this.onLogout();
     }
+  },
 
-    if (!pkg.get("isAvailable")) {
-      // Update our cart if a package has been dispatched
-      let undispatchedPkgs = pkg.get("item.undispatchedPackages");
-      if (undispatchedPkgs.length) {
-        this.removeItem(pkg.get("item"));
-        undispatchedPkgs
-          .map(p => p.toCartItem())
-          .forEach(ci => this.pushItem(ci));
-      } else {
-        this.setItemAvailability(cartItem, false); // none left
-      }
+  /**
+   * When the user logs in we
+   *    - fetch the user's cart items
+   *    - push any record he/she might have added as a guest before
+   *   logging in
+   */
+  onLogin() {
+    this.populate().then(() => {
+      this.persistLocalRecords();
+    });
+  },
+
+  /**
+   * When logging out, we clear the local data
+   */
+  onLogout() {
+    this.get("store")
+      .peekAll("cart_item")
+      .forEach(record => this.get("store").unloadRecord(record));
+  },
+
+  // ---- Control methods
+
+  /**
+   * Checkout the cart and fill the order
+   */
+  checkoutOrder(order) {
+    return this.POST(`/cart_items/checkout`, {
+      order_id: order.get("id")
+    }).then(data => {
+      this.get("store").pushPayload(data);
+      return this.refresh();
+    });
+  },
+
+  /**
+   * Fetches the user's cart items from the api
+   */
+  populate() {
+    if (!this.get("isLoggedIn")) {
+      return Ember.RSVP.resolve();
     }
+    return this.get("store").findAll("cart_item", { reload: true });
   },
 
-  setItemAvailability(cartItem, available) {
-    Ember.set(cartItem, "available", available);
-    this.pushItem(cartItem);
+  /**
+   * Clears the local data and re-fetches the data
+   */
+  refresh() {
+    this.get("store").unloadAll("cart_item");
+    return this.populate();
   },
 
-  content: computed(function() {
-    return JSON.parse(window.localStorage.getItem("cart") || "[]");
-  }),
+  /**
+   * Cart items that were added as guest are pushed to the db
+   */
+  persistLocalRecords() {
+    if (!this.get("isLoggedIn")) {
+      return Ember.RSVP.resolve();
+    }
 
-  cartItems: computed("[]", "@each.available", function() {
-    var content = this.get("content");
-    var allItems = [];
-    content.forEach(record => {
-      if (record.available) {
-        allItems.push(
-          this.get("store").peekRecord(record.modelType, record.id)
-        );
-      }
+    this.clearDuplicates();
```

Comment:
What is the reason to write this method `clearDuplicates` ? Why this Duplicates are getting added 🤔

Reply:
The case is : 

1. You log-in, and add items to your cart
2. You start an new session as a guest user (other device for example)
3. You add an item, as a guest, that is already in your cart
4. You log-in, the item is both in your cart, and stored locally as a "guest item"
5. We do a check to make sure that it's not there twice

The server would not allow the package to exist twice in the cart. But rather than having a pop-up error message I thought it would be best to clear those that are already added in the cart

Reply:
Is there any utility in lodash, which avoids duplicates while merging offline items with existing API cart ? Just a thought